### ISSUE TITLE
Runtime builds are too slow with the GPU locale model

### ISSUE THREAD

**Issue by e-kayrakli** [chapel-lang/chapel#23893]
GPU-based reduction support in the runtime relies on header-only template libraries provided by NVIDIA and CUDA. As we can't directly interop with C++, we use macro magic to stamp out all versions we care about, and put them in C functions. However, as the library is header-only, everything gets inlined and bloats the runtime build time significantly (several minutes), and likely the runtime binary size. Another dimension to the bloat is number of different offload architectures we compile for (currently 4).

Because that's what the API asks us to do, each function we use is called twice. We can probably put the calls to `statc inline`s in the libraries into a helper function, so that we can at least reduce the bloat by half. I am not sure if we can do much else.

**Comment by mppf** [comment]
I think this is an interesting motivating case for C++ interop.

Also, in case you are not aware of the pattern, you can instantiate templates in a .cpp file and declare them as instantiated templates in a .h file. It is a bit of a weird pattern but it is useful in some cases. Point is, you should not be required to inline these if you don't want to.

**Comment by e-kayrakli** [comment]
> Also, in case you are not aware of the pattern, you can instantiate templates in a .cpp file and declare them as instantiated templates in a .h file. It is a bit of a weird pattern but it is useful in some cases. Point is, you should not be required to inline these if you don't want to.

Yeah, I thought about it. What I have in mind is also a bit weird, but I think it is less so?

Right now what we have is:

```cpp
void reduce() {  // tons of stamped-out versions of this with macro magic
  static_inline_function_in_cub();  // this is a template function 
  static_inline_function_in_cub();  // but in a single `reduce` function, we use the same
                                    // instantiation twice
}
```

What I have in mind:

```cpp
void wrap() {
  static_inline_function_in_cub();
}

void reduce() {
  wrap();  // given that what we have today kills compilation performance,
  wrap();  // I am hoping that the compiler will not inline these calls
}
```

**Comment by e-kayrakli** [comment]
https://github.com/chapel-lang/chapel/pull/24006 will fix this.

Re inlining: it wasn't an issue. I implemented what I suggested above, but it didn't make any notable difference on my system. My theory is that the cost is creating instantiations, and not really using them even if they are `inline`. The compiler is probably doing the smart thing and throttling inlining, so us trying to avoid inlining doesn't make any difference.
